Tests fail with sympy >= 1.13.0 (PrintMethodNotImplementedError)

Majority of tests started to fail with sympy >= 1.13.0. Latests release, 1.13.1, still shows the same error. Issue is possibly related to sympy/sympy/issues/26663 or sympy/sympy/issues/26404. I haven't investigated, yet.

<details>
<summary> Example error message </summary>

```python
_________ ERROR at setup of test_IO_Rate.test_parameters_save_and_load _________
self = <ANNarchy.parser.Equation.Equation object at 0x7fc12c3399b0>
    def parse(self):
        "Main method called after creating the object."
        try:
            if self.type == 'ODE':
>               code = self.analyse_ODE(self.expression)
../BUILDROOT/usr/lib64/python3.13/site-packages/ANNarchy/parser/Equation.py:92: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../BUILDROOT/usr/lib64/python3.13/site-packages/ANNarchy/parser/Equation.py:242: in analyse_ODE
    return self.implicit(expression)
../BUILDROOT/usr/lib64/python3.13/site-packages/ANNarchy/parser/Equation.py:425: in implicit
    +  self.c_code(equation) + ';'
../BUILDROOT/usr/lib64/python3.13/site-packages/ANNarchy/parser/Equation.py:154: in c_code
    c_code = sp.ccode(
/usr/lib/python3.13/site-packages/sympy/printing/codeprinter.py:752: in ccode
    return c_code_printers[standard.lower()](settings).doprint(expr, assign_to)
/usr/lib/python3.13/site-packages/sympy/printing/codeprinter.py:172: in doprint
    lines = self._print(expr).splitlines()
/usr/lib/python3.13/site-packages/sympy/printing/printer.py:331: in _print
    return printmethod(expr, **kwargs)
/usr/lib/python3.13/site-packages/sympy/printing/c.py:142: in _meth_wrapper
    return meth(self, expr, **kwargs)
/usr/lib/python3.13/site-packages/sympy/printing/c.py:272: in _print_Mul
    return super()._print_Mul(expr, **kwargs)
/usr/lib/python3.13/site-packages/sympy/printing/codeprinter.py:565: in _print_Mul
    a_str = [self.parenthesize(x, prec) for x in a]
/usr/lib/python3.13/site-packages/sympy/printing/str.py:36: in parenthesize
    return "(%s)" % self._print(item)
/usr/lib/python3.13/site-packages/sympy/printing/printer.py:331: in _print
    return printmethod(expr, **kwargs)
/usr/lib/python3.13/site-packages/sympy/printing/str.py:57: in _print_Add
    t = self._print(term)
/usr/lib/python3.13/site-packages/sympy/printing/printer.py:331: in _print
    return printmethod(expr, **kwargs)
/usr/lib/python3.13/site-packages/sympy/printing/c.py:142: in _meth_wrapper
    return meth(self, expr, **kwargs)
/usr/lib/python3.13/site-packages/sympy/printing/c.py:272: in _print_Mul
    return super()._print_Mul(expr, **kwargs)
/usr/lib/python3.13/site-packages/sympy/printing/codeprinter.py:565: in _print_Mul
    a_str = [self.parenthesize(x, prec) for x in a]
/usr/lib/python3.13/site-packages/sympy/printing/str.py:36: in parenthesize
    return "(%s)" % self._print(item)
/usr/lib/python3.13/site-packages/sympy/printing/printer.py:331: in _print
    return printmethod(expr, **kwargs)
/usr/lib/python3.13/site-packages/sympy/printing/str.py:57: in _print_Add
    t = self._print(term)
/usr/lib/python3.13/site-packages/sympy/printing/printer.py:331: in _print
    return printmethod(expr, **kwargs)
/usr/lib/python3.13/site-packages/sympy/printing/codeprinter.py:461: in _print_Function
    return self._print_not_supported(expr)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <sympy.printing.c.C99CodePrinter object at 0x7fc12c2923f0>
expr = fa(__sum_exc__)
    def _print_not_supported(self, expr):
        if self._settings.get('strict', False):
>           raise PrintMethodNotImplementedError("Unsupported by %s: %s" % (str(type(self)), str(type(expr))) + \
                             "\nSet the printer option 'strict' to False in order to generate partially printed code.")
E           sympy.printing.codeprinter.PrintMethodNotImplementedError: Unsupported by <class 'sympy.printing.c.C99CodePrinter'>: fa
E           Set the printer option 'strict' to False in order to generate partially printed code.
/usr/lib/python3.13/site-packages/sympy/printing/codeprinter.py:582: PrintMethodNotImplementedError
During handling of the above exception, another exception occurred:
cls = <class 'Interface.test_IO.test_IO_Rate'>
    @classmethod
    def setUpClass(cls):
        """
        Compile the network for this test
        """
>       cls.network, cls.pop2, cls.proj2, cls.proj3 = define_rate_net()
tests/Interface/test_IO.py:63: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/Interface/networks.py:44: in define_rate_net
    pop2 = Population(name='pop2', neuron=DefaultNeuron, geometry=1)
../BUILDROOT/usr/lib64/python3.13/site-packages/ANNarchy/core/Population.py:101: in __init__
    self.neuron_type._analyse()
../BUILDROOT/usr/lib64/python3.13/site-packages/ANNarchy/core/Neuron.py:74: in _analyse
    self.description = analyse_neuron(self)
../BUILDROOT/usr/lib64/python3.13/site-packages/ANNarchy/parser/AnalyseNeuron.py:261: in analyse_neuron
    code = translator.parse()
../BUILDROOT/usr/lib64/python3.13/site-packages/ANNarchy/parser/Equation.py:103: in parse
    Messages._error('Parser: cannot analyse', self.expression)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
var_text = ('Parser: cannot analyse', 'tau * dmp/dt + mp = fa(__sum_exc__) +  neg(- fb(__sum_inh__)) + baseline + noise')
args = {}
text = 'Parser: cannot analyse tau * dmp/dt + mp = fa(__sum_exc__) +  neg(- fb(__sum_inh__)) + baseline + noise '
var = 'tau * dmp/dt + mp = fa(__sum_exc__) +  neg(- fb(__sum_inh__)) + baseline + noise'
exit = True
    def _error(*var_text, **args):
        """
        Prints an error message to standard out and exits.
    
        When passing exit=False, the program will not exit.
        """
        text = ''
        for var in var_text:
            text += str(var) + ' '
    
        exit = False
        if 'exit' in args.keys():
            if args['exit']:
                exit = True
        else:
            exit = True
    
        if exit:
>           raise ANNarchyException(text)
E           ANNarchy.intern.Messages.ANNarchyException: Parser: cannot analyse tau * dmp/dt + mp = fa(__sum_exc__) +  neg(- fb(__sum_inh__)) + baseline + noise
../BUILDROOT/usr/lib64/python3.13/site-packages/ANNarchy/intern/Messages.py:102: ANNarchyException
```

</details>

Full log available (for a limited time): https://kojipkgs.fedoraproject.org//work/tasks/5079/120765079/build.log

There are other errors logged, but I assume those are collateral of above error.

--- a/ANNarchy/parser/CoupledEquations.py
+++ b/ANNarchy/parser/CoupledEquations.py
@@ -9,7 +9,7 @@ from ANNarchy.intern import Messages
 from .Equation import Equation
 from .ParserTemplate import create_local_dict, user_functions
 
-from sympy import *
+import sympy as sp
 
 import re
 
@@ -56,6 +56,7 @@ class CoupledEquations(Equation):
 
     def parse(self):
         "Main method called after creating the object."
+        
         # Check if the numerical method is the same for all ODEs
         methods = []
         for var in self.variables:
@@ -99,29 +100,37 @@ class CoupledEquations(Equation):
             expression = expression.replace('_t_gradient_', '(_'+name+' - '+name+')')
             expression_list[name] = expression
 
-            new_var = Symbol('_'+name)
+            new_var = sp.Symbol('_'+name)
             self.local_dict['_'+name] = new_var
             new_vars[new_var] = name
 
         for name, expression in expression_list.items():
-            analysed = self.parse_expression(expression,
+            analysed = self.parse_expression(
+                expression,
                 local_dict = self.local_dict
             )
             equations[name] = analysed
 
         try:
-            solution = solve(list(equations.values()), list(new_vars.keys()))
+            solution = sp.solve(list(equations.values()), list(new_vars.keys()))
         except Exception as e:
             Messages._print(expression_list)
             Messages._error('The multiple ODEs can not be solved together using the implicit Euler method.')     
 
         for var, sol in solution.items():
-            # simplify the solution
-            sol  = collect( sol, self.local_dict['dt'])
+            # Simplify the solution
+            sol  = sp.collect( sol, self.local_dict['dt'])
+
+            # C++ code
+            try:
+                code = sp.ccode(sol, strict=False) 
+            except:
+                code = sp.ccode(sol) 
+
 
             # Generate the code
-            cpp_eq = get_global_config('precision') + ' _' + new_vars[var] + ' = ' + ccode(sol) + ';'
-            switch = ccode(self.local_dict[new_vars[var]] ) + ' = _' + new_vars[var] + ';'
+            cpp_eq = get_global_config('precision') + ' _' + new_vars[var] + ' = ' + code + ';'
+            switch = sp.ccode(self.local_dict[new_vars[var]] ) + ' = _' + new_vars[var] + ';'
 
             # Replace untouched variables with their original name
             for prev, new in sorted(list(self.untouched.items()), key = lambda key : len(key[0]), reverse=True):
@@ -157,7 +166,7 @@ class CoupledEquations(Equation):
             
             # Transform the gradient into a difference TODO: more robust..            
             expression = expression.replace('d'+name+'/dt', '_grad_var_'+name)
-            new_var = Symbol('_grad_var_'+name)
+            new_var = sp.Symbol('_grad_var_'+name)
             self.local_dict['_grad_var_'+name] = new_var
             
             expression_list[name] = expression
@@ -169,7 +178,7 @@ class CoupledEquations(Equation):
                 local_dict = self.local_dict
             )
             equations[name] = analysed
-            evaluations[name] = solve(analysed, self.local_dict['_grad_var_'+name])
+            evaluations[name] = sp.solve(analysed, self.local_dict['_grad_var_'+name])
 
         # Compute the k = f(x, t)
         ks = {}
@@ -181,7 +190,7 @@ class CoupledEquations(Equation):
         for name, val in self.local_dict.items():
             tmp_dict[name] = val
         for name, evaluation in evaluations.items():
-            tmp_dict[name] = Symbol('(' + ccode(self.local_dict[name]) + ' + 0.5 * dt * _k_' + name + ' )')
+            tmp_dict[name] = sp.Symbol('(' + sp.ccode(self.local_dict[name]) + ' + 0.5 * dt * _k_' + name + ' )')
 
         # Compute the new values _x_new = f(x + dt/2*_k)
         news = {}
@@ -189,13 +198,13 @@ class CoupledEquations(Equation):
             tmp_analysed = self.parse_expression(expression,
                 local_dict = tmp_dict
             )
-            solved = solve(tmp_analysed, self.local_dict['_grad_var_'+name])
+            solved = sp.solve(tmp_analysed, self.local_dict['_grad_var_'+name])
             news[name] = get_global_config('precision') + ' _' + name + ' = ' + self.c_code(solved[0]) + ';'
 
         # Compute the switches
         switches = {}
         for name, expression in expression_list.items():
-            switches[name] = ccode(self.local_dict[name]) + ' += dt * _' + name + ' ;'
+            switches[name] = sp.ccode(self.local_dict[name]) + ' += dt * _' + name + ' ;'
 
         # Store the generated code in the variables
         for name in self.names:
@@ -234,7 +243,7 @@ class CoupledEquations(Equation):
             expression = expression.replace(' ', '')
             # Transform the gradient into a difference TODO: more robust...
             expression = expression.replace('d'+name+'/dt', '_gradient_'+name)
-            self.local_dict['_gradient_'+name] = Symbol('_gradient_'+name)
+            self.local_dict['_gradient_'+name] = sp.Symbol('_gradient_'+name)
             expression_list[name] = expression
 
         for name, expression in expression_list.items():
@@ -242,7 +251,7 @@ class CoupledEquations(Equation):
                 local_dict = self.local_dict
             )
             equations[name] = analysed
-            evaluations[name] = solve(analysed, self.local_dict['_gradient_'+name])
+            evaluations[name] = sp.solve(analysed, self.local_dict['_gradient_'+name])
 
         # Compute the k1 = f(x, t)
         k1_dict = {}
@@ -255,7 +264,7 @@ class CoupledEquations(Equation):
         for name, val in self.local_dict.items():
             tmp_dict_k2[name] = val
         for name, evaluation in evaluations.items():
-            tmp_dict_k2[name] = Symbol('(' + ccode(self.local_dict[name]) + ' + 0.5 * dt * _k1_' + name + ' )')
+            tmp_dict_k2[name] = sp.Symbol('(' + sp.ccode(self.local_dict[name]) + ' + 0.5 * dt * _k1_' + name + ' )')
 
         # Compute the values _k2_x = f(x + dt/2*_k1)
         for name, expression in expression_list.items():
@@ -263,7 +272,7 @@ class CoupledEquations(Equation):
                 local_dict = tmp_dict_k2
             )
 
-            solved = solve(tmp_analysed, self.local_dict['_gradient_'+name])
+            solved = sp.solve(tmp_analysed, self.local_dict['_gradient_'+name])
             k2_dict[name] = get_global_config('precision') + ' _k2_' + name + ' = ' + self.c_code(solved[0]) + ';'
 
         # New dictionary replacing x by x+dt/2*k2)
@@ -272,7 +281,7 @@ class CoupledEquations(Equation):
         for name, val in self.local_dict.items():
             tmp_dict_k3[name] = val
         for name, evaluation in evaluations.items():
-            tmp_dict_k3[name] = Symbol('(' + ccode(self.local_dict[name]) + ' + 0.5 * dt *_k2_' + name + ' )')
+            tmp_dict_k3[name] = sp.Symbol('(' + sp.ccode(self.local_dict[name]) + ' + 0.5 * dt *_k2_' + name + ' )')
 
         # Compute the values _k3_x = f(x + dt/2*_k2_x)
         for name, expression in expression_list.items():
@@ -280,7 +289,7 @@ class CoupledEquations(Equation):
                 local_dict = tmp_dict_k3
             )
 
-            solved = solve(tmp_analysed, self.local_dict['_gradient_'+name])
+            solved = sp.solve(tmp_analysed, self.local_dict['_gradient_'+name])
             k3_dict[name] = get_global_config('precision') + ' _k3_' + name + ' = ' + self.c_code(solved[0]) + ';'
 
         # New dictionary replacing x by x+dt*k3)
@@ -289,7 +298,7 @@ class CoupledEquations(Equation):
         for name, val in self.local_dict.items():
             tmp_dict_k4[name] = val
         for name, evaluation in evaluations.items():
-            tmp_dict_k4[name] = Symbol('(' + ccode(self.local_dict[name]) + ' + dt*_k3_' + name + ' )')
+            tmp_dict_k4[name] = sp.Symbol('(' + sp.ccode(self.local_dict[name]) + ' + dt*_k3_' + name + ' )')
 
         # Compute the values _k4_x = f(x + dt*_k3_x)
         for name, expression in expression_list.items():
@@ -297,14 +306,14 @@ class CoupledEquations(Equation):
                 local_dict = tmp_dict_k4
             )
 
-            solved = solve(tmp_analysed, self.local_dict['_gradient_'+name])
+            solved = sp.solve(tmp_analysed, self.local_dict['_gradient_'+name])
             k4_dict[name] = get_global_config('precision') + ' _k4_' + name + ' = ' + self.c_code(solved[0]) + ';'
 
         # accumulate _k1 - _k4 within the switch step
         dt_code = "dt/6.0f" if _check_precision('float') else "dt/6.0"
         switches = {}
         for name, expression in expression_list.items():
-            switches[name] = ccode(self.local_dict[name]) + ' += '+dt_code+' * (_k1_'+name+' + (_k2_'+name+' + _k2_'+name+') + (_k3_'+name+' + _k3_'+name+') + _k4_'+name+');'
+            switches[name] = sp.ccode(self.local_dict[name]) + ' += '+dt_code+' * (_k1_'+name+' + (_k2_'+name+' + _k2_'+name+') + (_k3_'+name+' + _k3_'+name+') + _k4_'+name+');'
 
         # Store the generated code in the variables
         for name in self.names:

--- a/ANNarchy/parser/Equation.py
+++ b/ANNarchy/parser/Equation.py
@@ -151,11 +151,21 @@ class Equation(object):
         Changes to this method should be applied also
         on ANNarchy.parser.CoupledEquations.c_code() too.
         """
-        c_code = sp.ccode(
-            equation,
-            precision=8,
-            user_functions=self.user_functions
-        )
+        # The `strict` parameter has been introduced in sympy >= 1.13.
+        # It defaults to `True`.
+        try:
+            c_code = sp.ccode(
+                equation,
+                precision=8,
+                strict=False,
+                user_functions=self.user_functions
+            )
+        except TypeError:
+            c_code = sp.ccode(
+                equation,
+                precision=8,
+                user_functions=self.user_functions
+            )
 
         if get_global_config('precision')=="float":
             #
@@ -731,5 +741,3 @@ class Equation(object):
             if self.local_dict[att] in self.analysed.atoms():
                 deps.append(att)
         return deps
-
-

--- a/ANNarchy/parser/Function.py
+++ b/ANNarchy/parser/Function.py
@@ -93,8 +93,22 @@ class FunctionParser(object):
             Messages._print(expression)
             Messages._error('The function depends on unknown variables.')
 
-        return sp.ccode(eq, precision=8,
-            user_functions=self.user_functions)
+        # The `strict` parameter has been introduced in sympy >= 1.13.
+        # It defaults to `True`.
+        try:
+            c_code = sp.ccode(
+                eq,
+                precision=8,
+                strict=False,
+                user_functions=self.user_functions
+            )
+        except TypeError:
+            c_code = sp.ccode(
+                eq,
+                precision=8,
+                user_functions=self.user_functions
+            )
+        return c_code
 
     def dependencies(self):
         "For compatibility with Equation."

